### PR TITLE
Access control for Nova and Horizon

### LINKED ISSUE #434: Proper RBAC for Nova

### LINKED ISSUE #476: Add can-access-horizon and can-access-nova permissions

### PR DESCRIPTION
**Pull request by JoshOldenburg**
Fixes #434, fixes #476.

Essentially adds authorization requirements to all Nova pages with the existing permissions. This probably needs a decent amount of testing still so nothing slips through. Overall Nova access is limited to `access-nova`, which is given to admin and the officer roles, but after #505 that will be changed. Horizon is now limited by the `access-horizon` permission, which only admins get.

### CODE REVIEW

File: app/Nova/User.php [hunk 1]
```diff
@@ -190,10 +223,14 @@ public function cards(Request $request)
      */
     public function filters(Request $request)
     {
-        return [
-            new Filters\UserType,
+        $filters = [
             new Filters\UserActive,
         ];
+        if ($request->user()->hasRole('admin')) {
```
Comment (kberzinch) [inline app/Nova/User.php]: role check to discuss

File: app/Policies/AttendancePolicy.php [hunk 2]
```diff
@@ -67,7 +78,7 @@ public function delete(User $user, Attendance $attendance)
      */
     public function restore(User $user, Attendance $attendance)
     {
-        return false;
+        return $user->hasRole('admin');
```
Comment (kberzinch) [inline app/Policies/AttendancePolicy.php]: role check to discuss

File: app/Policies/DuesPackagePolicy.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Determine whether the user can view the dues package.
+     *
+     * @param  \App\User  $user
+     * @param  \App\DuesPackage  $duesPackage
+     * @return mixed
+     */
+    public function view(User $user, DuesPackage $duesPackage)
+    {
+        // Normal users have this, but Nova in general is limited by can-access-nova
```
Comment (kberzinch) [inline app/Policies/DuesPackagePolicy.php]: doc error: permission is `access-nova`

File: app/Policies/DuesPackagePolicy.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Determine whether the user can restore the dues package.
+     *
+     * @param  \App\User  $user
+     * @param  \App\DuesPackage  $duesPackage
+     * @return mixed
+     */
+    public function restore(User $user, DuesPackage $duesPackage)
+    {
+        return $user->hasRole('admin');
```
Comment (kberzinch) [inline app/Policies/DuesPackagePolicy.php]: role check to discuss

File: app/Policies/EventPolicy.php [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Determine whether the user can view the event.
+     *
+     * @param  \App\User  $user
+     * @param  \App\Event  $event
+     * @return mixed
+     */
+    public function view(User $user, Event $event)
+    {
+        // Normal users have this, but Nova in general is limited by can-access-nova
```
Comment (kberzinch) [inline app/Policies/EventPolicy.php]: doc error: permission is `access-nova`

File: app/Policies/EventPolicy.php [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    /**
+     * Determine whether the user can restore the event.
+     *
+     * @param  \App\User  $user
+     * @param  \App\Event  $event
+     * @return mixed
+     */
+    public function restore(User $user, Event $event)
+    {
+        return $user->hasRole('admin');
```
Comment (kberzinch) [inline app/Policies/EventPolicy.php]: role check to discuss

File: app/Policies/NotificationTemplatePolicy.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+    /**
+     * Determine whether the user can view the notification template.
+     *
+     * @param  \App\User  $user
+     * @param  \App\NotificationTemplate  $notificationTemplate
+     * @return mixed
+     */
+    public function view(User $user, NotificationTemplate $notificationTemplate)
+    {
+        return $user->hasRole('admin');
```
Comment (kberzinch) [inline app/Policies/NotificationTemplatePolicy.php]: role checks everywhere to discuss

File: app/Policies/TeamPolicy.php [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+    public function attachUser(User $user, Team $team, User $userResource)
+    {
+        if ($team->members->contains('id', $userResource->id)) {
+            return false;
+        }
+        if (! $team->visible && $user->cant('read-teams-hidden')) {
+            return false;
+        }
+
+        return $user->can('update-teams-membership');
```
Comment (kberzinch) [inline app/Policies/TeamPolicy.php]: add "project manager" flag on teams?

File: app/Policies/UserPolicy.php [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    /**
+     * Determine whether the user can restore the user.
+     *
+     * @param  \App\User  $user
+     * @param  \App\User  $user
+     * @return mixed
+     */
+    public function restore(User $user, User $userResource)
+    {
+        return $user->hasRole('admin');
```
Comment (kberzinch) [inline app/Policies/UserPolicy.php]: role check to discuss

File: app/Policies/UserPolicy.php [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+    public function attachTeam(User $user, User $userResource, Team $team)
+    {
+        if ($team->members->contains('id', $userResource->id)) {
+            return false;
+        }
+        if (! $team->visible && $user->cant('read-teams-hidden')) {
+            return false;
+        }
+
+        return $user->can('update-teams-membership');
```
Comment (kberzinch) [inline app/Policies/UserPolicy.php]: PM tag

File: app/Providers/NovaServiceProvider.php [hunk 11]
```diff
@@ -78,8 +78,12 @@ protected function cards()
     public function tools()
     {
         return [
-            new \Vyuldashev\NovaPermission\NovaPermissionTool(),
-            new AttendanceReport(),
+            (new \Vyuldashev\NovaPermission\NovaPermissionTool())->canSee(function ($request) {
+                return $request->user()->hasRole('admin');
```
Comment (kberzinch) [inline app/Providers/NovaServiceProvider.php]: role check

Comment (JoshOldenburg) [reply app/Providers/NovaServiceProvider.php]: Approved in Slack

File: app/Policies/AttendancePolicy.php (hunk 2 above)
Comment (JoshOldenburg) [reply app/Policies/AttendancePolicy.php]: Leaving this as admin since users can't create attendance in Nova yet

**Pull request merged**